### ISSUE TITLE
Don't convert arguments based on default value type if there is type hint

### ISSUE THREAD

**Issue by pekkaklarck** [robotframework/robotframework#4881]
Currently if an argument has a type hint and a default value like `arg: int = None`, we first try conversion based on the type hint and if it fails we also try conversion based on the default value. For example, a keyword with the following argument specification could be called like
```
Keyword    42
Keyword    None
```
and both usages would work.

This doesn't make much sense, because the type hint not containing the default value type is actually a typing error. For example, the previous example should be `arg: Union[int, None] = None` (or `arg: int|None = None` when using Python 3.10 or newer). This logic is also bit more complicated to explain and also makes the conversion code a bit more complicated.

I believe we should change the conversion logic so that if an argument has a type hint, we do conversion solely based on it, regardless does the argument have a default value or not. If argument doesn't have a type hint, we should do conversion based on the possible default value, though.

In practice this change would mean that a keyword with an argument like `arg: int = None` could be called with an argument that can be converted to an integer like
```
Keyword    42
```
and using
```
Keyword    None
```
fails. This is obviously backwards incompatible, but, as already mentioned, the typing is wrong and should be fixed.

**Comment by pekkaklarck** [comment]
I don't expect there to be much code with totally invalid typing like `arg: int = 'bad`. Because it's not generally recommended to use mutable default values like `arg: list = []`, it is possible that code like `arg: list = None` exists, though. Even Python itself used to consider this usage ok until [Python 3.11](https://docs.python.org/3/library/typing.html#typing.get_type_hints) and treated it the same way as `arg: Optional[list] = None` which in turn is just a shorthand for `arg: Uniont[list, None] = None`.

**Comment by pekkaklarck** [comment]
After thinking about this a bit more, and after changing the code accordingly and running all our acceptance tests, I don't think this is a good idea right now.

As I mentioned in the previous comment, Python itself handles the `arg: list = None` case differently depending on the version. With Python 3.10 and earlier, it is the same as `arg: Union[list, None] = None`, which means that our argument conversion code does `None` conversion already based on the type hint. With Python 3.11 it stays as `arg: list = None`, but because our conversion code looks also at the default value, `None` conversion is done as well. In other words, if we stop looking at the default value, the decision to do `None` conversion depends on the Python version. That would be confusing to users and also would make our tests version specific.

I still believe we should do conversion only based on the type hint if it exists, but I think it's better to wait until we have Python 3.11 as the minimum Python version. Then the conversion behavior won't change between versions and it ought to be more well known that the default value type needs to be specified explicitly in type hints.

I close this issue now, but I add a comment to the code about reconsidering this in the future. I'll also add a note to the User Guide that this behavior may change.

**Comment by pekkaklarck** [comment]
Let's do this in RF 8.0. It will simplify argument conversion code and also help with #5646.

We still need to decide what to do with Python 3.11 handling `arg: list = None` differently than earlier versions. Python 3.10 has already [reached its end-of-life](https://endoflife.date/python) by the time RF 8.0 will be released, so we can make Python 3.11 our minimum version. If that is considered too early, and this is the only case where differences between Python 3.10 and 3.11 matter to us, Python 3.10 having different semantics compared to never version ought to be fine.
